<a href="https://colab.research.google.com/github/shuvechchha/Breast-Cancer-Prediction-Model/blob/main/Breast_Cancer_Prediction_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 1

Project 1 is about using cross-validation to explore hyperparameter options for two different types of models. We are using a random forest model and a k-nearest neighbors model.

We are using another dataset with a long history in machine learning, a dataset from the Univerisity of Wisconsin on breast cancer.

Be sure to mount your drive so that you can access the data set.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_validate
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler

In [ ]:
# Read in the dataset and specify the id column to be row indices
bcwd = pd.read_csv('/content/drive/My Drive/HDS/Datasets/breast_cancer_wisconsin_diagnostic.csv', index_col='id')
# Recode the classes with numbers
bcwd['diagnosis'].replace('M',1, inplace=True)
bcwd['diagnosis'].replace('B',0, inplace=True)
bcwd.head()

diagnosis  radius_mean  ...  symmetry_worst  fractal_dimension_worst
id                                ...                                         
842302            1        17.99  ...          0.4601                  0.11890
842517            1        20.57  ...          0.2750                  0.08902
84300903          1        19.69  ...          0.3613                  0.08758
84348301          1        11.42  ...          0.6638                  0.17300
84358402          1        20.29  ...          0.2364                  0.07678

[5 rows x 31 columns]

In [ ]:
# Create training and test sets
# Using a seed for the random number generator to ensure the split is the same every time
# for the purposes of this project
training_set, test_set = train_test_split(bcwd, test_size=0.20, random_state=42)

# Use all of the features, excluding diagnosis at the beginning
training_features = training_set.iloc[:,1:]
# Targets are at the end
training_targets = training_set.iloc[:,0]

# Pull out the same features and targets from test set
testing_features = test_set.iloc[:,1:]
testing_targets = test_set.iloc[:,0]

# Use this list of measurements to score your cross-validation results
scoring = ['accuracy','precision', 'recall', 'f1']

# Part One

In this part we will train a series of random forest models using cross-validation. We will perform what is called a grid search of two different hyperparameters. A grid search explores different combinations of values across multiple hyperparamters. 

For our purposes, we will explore max_features and max_depth. Max_features specifies how many features each tree has access too. Max_depth limits the growth of the tree. We will explore two values (small and large) for max_features, 4 and 20, and two values of max_depth, 4 and 10. This means a total of 4 models will be created. 

The first model with max_features=4 and max_depth=4 has code ready to go. You will need to create the other three models that have different pairings of hyperparameter values. Finally, pick your model with the highest test_accuracy and evaluate it on the test set (Code is given for this, but you must set the hyperparameters based on your results).


In [ ]:
# Sample code of a random forest cross-validation run

# Create the random forest model with max_features=4 and max_depth=4
rf_model1 = RandomForestClassifier(criterion='entropy', max_features=4, max_depth=4, random_state=3)

# Cross validation calculating multiple metrics
rf_model1.fit(training_features, training_targets)
scores1 = cross_validate(rf_model1, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results1 = pd.DataFrame(scores1)
results1.mean()

fit_time           0.198069
score_time         0.014280
test_accuracy      0.956044
train_accuracy     0.992857
test_precision     0.952086
train_precision    1.000000
test_recall        0.929055
train_recall       0.980773
test_f1            0.939615
train_f1           0.990290
dtype: float64

In [ ]:
# Create the random forest model with max_features=4 and max_depth=10

rf_model2 = RandomForestClassifier(criterion='entropy', max_features=4, max_depth=10, random_state=3)

rf_model2.fit(training_features, training_targets)
scores2 = cross_validate(rf_model2, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

results2 = pd.DataFrame(scores2)
results2.mean()

fit_time           0.205838
score_time         0.012627
test_accuracy      0.958242
train_accuracy     1.000000
test_precision     0.963473
train_precision    1.000000
test_recall        0.923173
train_recall       1.000000
test_f1            0.942157
train_f1           1.000000
dtype: float64

In [ ]:
# Create the random forest model with max_features=20 and max_depth=4
rf_model3 = RandomForestClassifier(criterion='entropy', max_features=20, max_depth=4, random_state=3)

rf_model3.fit(training_features, training_targets)
scores3 = cross_validate(rf_model3, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

results3 = pd.DataFrame(scores3)
results3.mean()

fit_time           0.410085
score_time         0.011585
test_accuracy      0.960440
train_accuracy     0.995055
test_precision     0.964221
train_precision    1.000000
test_recall        0.929055
train_recall       0.986688
test_f1            0.945502
train_f1           0.993292
dtype: float64

In [ ]:
# Create the random forest model with max_features=20 and max_depth=10
rf_model4 = RandomForestClassifier(criterion='entropy', max_features=20, max_depth=10, random_state=3)

rf_model4.fit(training_features, training_targets)
scores4 = cross_validate(rf_model4, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

results4 = pd.DataFrame(scores4)
results4.mean()

fit_time           0.418800
score_time         0.013035
test_accuracy      0.958242
train_accuracy     1.000000
test_precision     0.957971
train_precision    1.000000
test_recall        0.929055
train_recall       1.000000
test_f1            0.942521
train_f1           1.000000
dtype: float64

In [ ]:
# Sample code evaluating an rf model on the test set
# Customize it as appropriate
rf_model1 = RandomForestClassifier(criterion='entropy', max_features=4, max_depth=4)
rf_model1.fit(training_features, training_targets)

# The score method is quick way of getting the accuracy on the testing data
rf_model1.score(testing_features, testing_targets)

0.9649122807017544

In [ ]:
rf_model2 = RandomForestClassifier(criterion='entropy', max_features=4, max_depth=10)
rf_model2.fit(training_features, training_targets)
rf_model2.score(testing_features, testing_targets)

0.9649122807017544

In [ ]:
rf_model3 = RandomForestClassifier(criterion='entropy', max_features=20, max_depth=4)
rf_model3.fit(training_features, training_targets)
rf_model3.score(testing_features, testing_targets)

0.9649122807017544

In [ ]:
rf_model4 = RandomForestClassifier(criterion='entropy', max_features=20, max_depth=10)
rf_model4.fit(training_features, training_targets)
rf_model4.score(testing_features, testing_targets)

0.956140350877193

## Questions on Part One

1) Which model had the highest accuracy test_accuracy score?

2) Which model had the highest test_f1 score?

3) Which model had the highest fit_time? Why do you think this is?

4) Which model had the lowest fit_time? Why do you think this is?

5) Does the model with the highest fit_time have the best f1 score?

6) Did you see any evidence of over-fitting from the training accuracy scores?

7) How did your best model in terms of test_accuracy score do when evaluated on the testing set using the score method? How different were the results from the test_accuracy score during cross-validation? 


1.   The third model with max_features=20 and max_depth=4 has the highest accuracy score of 0.960440.
2.  The third model with max_features=20 and max_depth=4 has the highest test_f1 score of 0.945502. 
3. The fourth model with max_features=20 and max_depth=10 has the highest fit_time score of 0.418800. Since this model considers the highest number of features and depth, it must take longer time to fit. 
4.   The first model with max_features=4 and max_depth=4 has the lowest fit_time score of 0.198069. Since this model considers the lowest number of features and depth, it might have the lowest fit_time. 
5.   No, the model with the highest fit_time score does not have the best f1 score. 
6.  There is no significant evidence of overfitting from the trianing accuracy scores.   
7.   When the best model on accuracy was tested on the testing set using the score method, it received a higher score of 0.9649122. Except for the fourth model with max_features=20 and max_depth=10, which has a test accuracy score of 0.956140, all of the models have the same test accuracy score of 0.9649122. When tested on the testing set using the score method, the fourth model scored lower on test accuracy, whilst other models scored higher.










# Part Two
In part two we will explore using k-nearest neighbors algorithm. First, we will investigate whether scaling the data improves results. Second, we ill use cross-validation to set the k value of our model. 

Code is given to create and evaluate a k-nearest neighbors model with K=3 on the original data. Code is then given to scale the data. You need to then recreate the knn model with K=3 and evalute its performance on the scaled data. Then create knn models with K=7 and K=12 and evaluate them. Finally, pick your model with the highest test_accuracy and evaluate it on the test set (Code is given for this, but you must set the correct K value).


In [ ]:
# Sample code of a KNN cross-validation run

# Create KNN model with n_neighbors=3, train on original data
knn_model1 = KNeighborsClassifier(n_neighbors=3)

# Cross validation calculating multiple metrics
knn_model1.fit(training_features, training_targets)
scores1 = cross_validate(knn_model, training_features, training_targets, scoring=scoring, return_train_score=True, cv=5)

# Average the runs
results1 = pd.DataFrame(scores1)
results1.mean()

fit_time           0.002319
score_time         0.007867
test_accuracy      0.920879
train_accuracy     0.948901
test_precision     0.934580
train_precision    0.964923
test_recall        0.846702
train_recall       0.895000
test_f1            0.886192
train_f1           0.928605
dtype: float64

In [ ]:
# Scale the features to have a mean of 0 and a standard deviation of 1
feature_scaler = StandardScaler().fit(training_features)
training_features_scaled = feature_scaler.transform(training_features)
testing_features_scaled = feature_scaler.transform(testing_features)

In [ ]:
# Create KNN model with n_neighbors=3, train on scaled data
knn_model1 = KNeighborsClassifier(n_neighbors=3)
knn_model1.fit(training_features_scaled, training_targets)
scores1 = cross_validate(knn_model1, training_features_scaled, training_targets, scoring=scoring, return_train_score=True, cv=5)

results1 = pd.DataFrame(scores1)
results1.mean()

fit_time           0.001718
score_time         0.009085
test_accuracy      0.960440
train_accuracy     0.982418
test_precision     0.974531
train_precision    0.996947
test_recall        0.917291
train_recall       0.955621
test_f1            0.944211
train_f1           0.975814
dtype: float64

In [ ]:
# Create KNN model with n_neighbors=7, train on scaled data
knn_model2 = KNeighborsClassifier(n_neighbors=7)
knn_model2.fit(training_features_scaled, training_targets)
scores2 = cross_validate(knn_model2, training_features_scaled, training_targets, scoring=scoring, return_train_score=True, cv=5)

results2 = pd.DataFrame(scores2)
results2.mean()

fit_time           0.001508
score_time         0.008172
test_accuracy      0.960440
train_accuracy     0.971429
test_precision     0.981339
train_precision    0.998374
test_recall        0.911586
train_recall       0.924575
test_f1            0.943823
train_f1           0.959980
dtype: float64

In [ ]:
# Create KNN model with n_neighbors=12, train on scaled data
knn_model3 = KNeighborsClassifier(n_neighbors=12)
knn_model3.fit(training_features_scaled, training_targets)
scores3 = cross_validate(knn_model3, training_features_scaled, training_targets, scoring=scoring, return_train_score=True, cv=5)

results3 = pd.DataFrame(scores3)
results3.mean()

fit_time           0.001645
score_time         0.009588
test_accuracy      0.953846
train_accuracy     0.963187
test_precision     0.993548
train_precision    0.998319
test_recall        0.881818
train_recall       0.902364
test_f1            0.933414
train_f1           0.947875
dtype: float64

In [ ]:
# Sample code evaluating a knn model on the scaled test set
# Customize it as appropriate
knn_model1 = KNeighborsClassifier(n_neighbors=3)
knn_model1.fit(training_features_scaled, training_targets)

# The score method is quick way of getting the accuracy on the testing data
knn_model1.score(testing_features_scaled, testing_targets)

0.9473684210526315

In [ ]:
knn_model2 = KNeighborsClassifier(n_neighbors=7)
knn_model2.fit(training_features_scaled, training_targets)
knn_model2.score(testing_features_scaled, testing_targets)

0.9473684210526315

In [ ]:
knn_model3 = KNeighborsClassifier(n_neighbors=12)
knn_model3.fit(training_features_scaled, training_targets)
knn_model3.score(testing_features_scaled, testing_targets)

0.956140350877193

## Questions for Part Two

1) Did the results from KNN model change when you switched to scaled features?

2) Which KNN model had the highest test_accuracy? Did it also have the highest test_f1 score? Is precision or recall higher and what does that say about the model's performance?

3) How did your best model in terms of test_accuracy do when evaluated on the testing set? Did the accuracy go up or down and by how much?

4) How does your best KNN model compare to your best Random Forest model in terms of accuracy on the testing set?

5) How do fit_times compare between the KNN models and the Random Forest models?

1.   The results from the KNN model did change when switched to scaled features. 
2.   The first model with K=3 and the second model with K=7 had the highest test accuracy of 0.960440. The first model with K=3 had the highest test_f1 score of 0.944211 and the highest test recall of 0.917291. The third model with K=12 had the highest test precision of 0.993548. Test precision is higher than test recall in all three models which means that they have higher true positive rates than false positive rates. 
3.   The test accuracy of both the first and second models decreased from 0.960440 to 0.947368 which is a difference of 0.013072, when assessed on the testing set.
4.  On the testing set, the best Random Forest model outperformed the best KNN model. 
5.  The Random Forest models outperformed the KNN models in terms of fit  times. The fit times for the Random Forest models ranged from 0.198069 to 0.418800, whereas the fit times for the KNN models ranged from 0.001508 to 0.002319.






